### PR TITLE
increase MAX_ASSETS_FOR_BUY_EXECUTION

### PR DESCRIPTION
**Pull request by xlc**
Partially addresses #1638

Still need a better solution to allow devs to have better control of this.

### CODE REVIEW

File: polkadot/xcm/xcm-builder/src/barriers.rs [hunk 1]
```diff
@@ -52,7 +52,7 @@ impl ShouldExecute for TakeWeightCredit {
 	}
 }
 
-const MAX_ASSETS_FOR_BUY_EXECUTION: usize = 1;
+const MAX_ASSETS_FOR_BUY_EXECUTION: usize = 5;
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/barriers.rs:55]: just wonder (maybe I am missing something), if it is good idea to expose and make it runtime-configurable like:
```
pub struct AllowTopLevelPaidExecutionFrom<T, MaxAssetsForBuyExecution>(PhantomData<(T, MaxAssetsForBuyExecution)>);

impl<
    T: Contains<MultiLocation>,
    MaxAssetsForBuyExecution: Get<u8>
> ShouldExecute for AllowTopLevelPaidExecutionFrom<T> {
```

**Review by acatangiu — APPROVED** [review]

**Comment by albertov19** [comment]
@KiChjang any comments here? This is needed for Foreign Assets in AssetHub for example. That is, if we want to register GLMR in AssetHub, we need to be able to pay with DOT XCM execution, or I'm missing something?

File: polkadot/xcm/xcm-builder/src/barriers.rs (hunk 1 above)
Comment (kianenigma) [reply polkadot/xcm/xcm-builder/src/barriers.rs:55]: Let's already do this if this needs to be 5 only in asset-hub?

Comment (xlc) [reply polkadot/xcm/xcm-builder/src/barriers.rs:55]: well, the update should really be universal. all the chains that supports multiple assets needs this.

**Comment by joepetrowski** [comment]
This might be needed for backwards compatibility with some programs existing on parachains, but I think the name of the variable makes the name clear: Max assets **for Buy Execution**. It is _not_ max assets in holding.

As the Barrier changes show [where it was added](https://github.com/paritytech/polkadot-sdk/commit/26be0727e5aaa5ed569b29bd9a6b043899a5617b), it's looking for the asset withdrawal before `BuyExecution`. I think the correct way to address the linked issue is to use the following pattern when you want to deposit more than one asset:

```
WithdrawAsset { /* 1 asset for fees */ }
BuyExecution { .. }
WithdrawAsset { /* other assets to use in the program */ }
...
```

File: polkadot/xcm/xcm-builder/src/barriers.rs (hunk 1 above)
Comment (joepetrowski) [reply polkadot/xcm/xcm-builder/src/barriers.rs:55]: No, not that support multiple asset. That support _buying execution_ in multiple assets.

Asset Hub does not support this.

Comment (joepetrowski) [reply polkadot/xcm/xcm-builder/src/barriers.rs:55]: You can see how this works now, using a [Take First Asset Trader](https://github.com/paritytech/polkadot-sdk/blob/bdbe982/cumulus/primitives/utility/src/lib.rs#L120). It (conveniently) allows you to put `MAX_ASSETS` into holding, but only takes the first one for `BuyExecution`. So this is still the same behavior; it allows 1 asset for `BuyExecution`.

Again, I know this PR might solve the short-term problem for existing use cases, but I think the intent of this `const` is clear and that a chain supporting multiple assets does not mean it supports multiple assets simultaneously for weight payment.

Comment (KiChjang) [reply polkadot/xcm/xcm-builder/src/barriers.rs:55]:
```suggestion
const MAX_ASSETS_FOR_BUY_EXECUTION: usize = 2;
```

**Review by franciscoaguirre — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Review by KiChjang — APPROVED** [review]

**Comment by acatangiu** [comment]
`continuous-integration/gitlab-check-runtime-migration-rococo` started failing for all PRs - unrelated to this

**Comment by bkchr** [comment]
> `continuous-integration/gitlab-check-runtime-migration-rococo` started failing for all PRs - unrelated to this

Is this already tracked? Looks like a migration is failing.

**Pull request merged**